packages/flutter/lib/src/widgets/magnifier.dart

```diff
@@ -0,0 +1,545 @@
+/// entirely covered by the shadow.
+///
+/// The negative space of the donut is clipped out (the donut hole, outside the donut).
+/// The donut hole is cut out exactly like the shape of the magnifier.
+class _DonutClip extends CustomClipper<Path> {
+  _DonutClip({required this.shape, required this.spreadRadius});
+
+  final double spreadRadius;
+  final ShapeBorder shape;
+
+  @override
+  Path getClip(Size size) {
+    final Path path = Path();
+    final Rect rect = Offset.zero & size;
+
+    path.fillType = PathFillType.evenOdd;
+    path.addPath(shape.getOuterPath(rect.inflate(spreadRadius)), Offset.zero);
+    path.addPath(shape.getInnerPath(rect), Offset.zero);
+    return path;
+  }
+
+  @override
+  bool shouldReclip(_DonutClip oldClipper) => oldClipper.shape != shape;
+}
+
+class _Magnifier extends SingleChildRenderObjectWidget {
+  /// Construct a [_Magnifier].
+  const _Magnifier({
+      super.child,
+      required this.shape,
+      this.magnificationScale = 1,
+      this.focalPoint = Offset.zero,
+  });
+
+  /// [focalPoint] is the area the center of the
+  /// [_Magnifier] points to, relative to the center of the magnifier.
+  /// If left as [Offset.zero], the magnifier will magnify whatever is directly
+  /// below it.
+  final Offset focalPoint;
+
+  /// The scale of the magnification.
+  ///
+  /// A [magnificationScale] of 1 means that the content magi
+  final double magnificationScale;
+
+  /// The shape of the magnifier is dictated by [clip], which clips
+  /// the magnifier to the shape. If null, the shape will be rectangular.
+  final ShapeBorder shape;
+
+  @override
+  RenderObject createRenderObject(BuildContext context) {
+    return _RenderMagnification(focalPoint, magnificationScale, shape);
+  }
+
+  @override
+  void updateRenderObject(
+      BuildContext context, covariant RenderProxyBox renderObject) {
+    (renderObject as _RenderMagnification)
+      ..focalPoint = focalPoint
+      ..shape = shape
+      ..magnificationScale = magnificationScale;
+  }
+}
+
+class _RenderMagnification extends RenderProxyBox {
+  _RenderMagnification(
+    this._focalPoint,
+    this._magnificationScale,
+    this._shape, {
+    RenderBox? child,
+  }) : super(child);
+
+  Offset get focalPoint => _focalPoint;
+  Offset _focalPoint;
+  set focalPoint(Offset value) {
+    if (_focalPoint == value) {
+      return;
+    }
+    _focalPoint = value;
+    markNeedsLayout();
+  }
+
+  double get magnificationScale => _magnificationScale;
+  double _magnificationScale;
+  set magnificationScale(double value) {
+    if (_magnificationScale == value) {
+      return;
+    }
+    _magnificationScale = value;
+    markNeedsLayout();
+  }
+
+  ShapeBorder get shape => _shape;
+  ShapeBorder _shape;
+  set shape(ShapeBorder value) {
+    if (_shape == value) {
+      return;
+    }
+    _shape = value;
+    markNeedsLayout();
+  }
+
+  @override
+  _MagnificationLayer? get layer => super.layer as _MagnificationLayer?;
```

Comment:
nit: you could just create a local variable in paint and do the down-casting there.

Reply:
I think this is idiomatic. RenderBackdropFilterLayer:

```dart
  @override
  BackdropFilterLayer? get layer => super.layer as BackdropFilterLayer?;
```

as well as ShaderMaskLayer, FollowerLayer use this form of casting. I also think it makes more sense to the end reader; the layer is a mag layer, not a container layer.

feel free to veto